internal/modprobe/modprobe_linux.go

```diff
@@ -0,0 +1,110 @@
+// Package modprobe attempts to load kernel modules. It may have more success
+// than simply running "modprobe", particularly for docker-in-docker.
+package modprobe
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/containerd/log"
+	"golang.org/x/sys/unix"
+)
+
+// LoadModules attempts to load kernel modules, if necessary.
+//
+// isLoaded must be a function that checks whether the modules are loaded. It may
+// be called multiple times. isLoaded must return an error to indicate that the
+// modules still need to be loaded, otherwise nil.
+//
+// For each method of loading modules, LoadModules will attempt the load for each
+// of modNames, then it will call isLoaded to check the result - moving on to try
+// the next method if needed, and there is one.
+//
+// The returned error is the result of the final call to isLoaded.
+func LoadModules(ctx context.Context, isLoaded func() error, modNames ...string) error {
+	if isLoaded() == nil {
+		log.G(ctx).WithFields(log.Fields{
+			"modules": strings.Join(modNames, ","),
```

Comment:
I was curious if we should let `modNames` as-is, or join them; `log.Fields` should  be able to handle either, but when using "text-format" logs would print them using the go string formatter (`[module-a module-b module-c]` which could indeed be slightly confusing (no commas). OTOH, when using `--log-format=json` it would preserve them as structured info; https://go.dev/play/p/TvOYyx1ze6R

```go
package main

import (
	"context"
	"strings"

	"github.com/containerd/log"
)

func main() {
	modNames := []string{"module-a", "module-b", "module-c"}

	_ = log.SetFormat(log.TextFormat)
	log.G(context.TODO()).WithFields(log.Fields{"modules": modNames}).Info("log as slice")
	log.G(context.TODO()).WithFields(log.Fields{"modules": strings.Join(modNames, ",")}).Info("log as string")
	_ = log.SetFormat(log.JSONFormat)
	log.G(context.TODO()).WithFields(log.Fields{"modules": modNames}).Info("log as slice")
	log.G(context.TODO()).WithFields(log.Fields{"modules": strings.Join(modNames, ",")}).Info("log as string")
}
```

```console
time="2009-11-10T23:00:00.000000000Z" level=info msg="log as slice" modules="[module-a module-b module-c]"
time="2009-11-10T23:00:00.000000000Z" level=info msg="log as string" modules="module-a,module-b,module-c"
```

```json
{"level":"info","modules":["module-a","module-b","module-c"],"msg":"log as slice","time":"2009-11-10T23:00:00.000000000Z"}
{"level":"info","modules":"module-a,module-b,module-c","msg":"log as string","time":"2009-11-10T23:00:00.000000000Z"}
```

Reply:
Oh, not sure - I joined the strings to make the log entry easier to read, but hadn't considered JSON output ... if it's better not to join them, I can do that?

Reply:
It's definitely not a hard blocker; mostly in light of "simplify if we don't _really_ need it" (looking at the above output, I think it's still ~ readable), and wondering if having it as actually structured info  (in JSON) would be beneficial.